datalad/core/distributed/clone.py

```diff
@@ -583,8 +583,11 @@ def postclonecfg_ria(ds, props):
     # get probe this RIA store when obtaining subdatasets
     ds.config.set(
         # we use the label 'origin' for this candidate in order to not have to
-        # generate a complicated name from the actual source specification
-        'datalad.get.subdataset-source-candidate-origin',
+        # generate a complicated name from the actual source specification.
+        # we pick a priority of 20 to sort it before datalad's default candidates
+        # for non-RIA URLs, because they prioritize hierarchical layouts that
+        # cannot be found in a RIA store
+        'datalad.get.subdataset-source-candidate-20origin',
```

Comment:
I wondered if it might make sense to separate out priority from the variable: i.e. introduce e.g.`datalad.get.subdataset-source-candidate-origin-priority` which if not set would get `20`.  That would 
- make existing clones work without any changes
- make it more explicit
- internal implementation could do whatever it does now (I would have preferred to have priority an explicit variable/tuple element to avoid all the splitting etc)  and could be refactored later if desired/needed

Reply:
If a second variable is introduce, one would also need to make sure that they stay in sync across all sources we read config from. I do not know how to achieve that, and how to test whether I have achieved that. It is perfectly possible that `datalad.get.subdataset-source-candidate-origin-priority` comes from user config and `datalad.get.subdataset-source-candidate-origin` does not.

I am not much concerned about existing clones -- I have all of them in front of me.

Reply:
I addressed the topic of the messy internal implementation in 05e4d8f600cac4fcb999c88e89deaaa184e1c1b9 and it also switched to a three-digit priority setup.

Reply:
> It is perfectly possible that datalad.get.subdataset-source-candidate-origin-priority comes from user config and datalad.get.subdataset-source-candidate-origin does not.

Wouldn't that be a feature - I could boost or penalize priority of all ria stores with one config setting?

Reply:
Besides I would follow annex and call it cost and not priority (ambiguous - is higher number a higher priority? More obvious with cost, and is taking the one with lower cost first).

Reply:
And in current approach it would be easy only to raise priority (lower cost) but not to lower it since the original variable would still be available somewhere with higher cost?

Reply:
> Wouldn't that be a feature - I could boost or penalize priority of all ria stores with one config setting?

That may be, but a feature that I do not comprehend enough to be the one implementing it, and there wasn't a use case for it either.

> Besides I would follow annex and call it cost and not priority (ambiguous - is higher number a higher priority? More obvious with cost, and is taking the one with lower cost first).

I can do this, if needed (but it may not be enough, see below).

> And in current approach it would be easy only to raise priority (lower cost) but not to lower it since the original variable would still be available somewhere with higher cost?

There is no concept of raising or lowering, there are just candidates with (default) priorities.

I suspect that you want to aim for a bigger solution. If you insist, I can obscure the fact that the sorting can be used to achieve a prioritizing and just talk about sorting. In this case the space would be open for the feature that you have in mind. I just want to get a handle on gh-4613, have that quickly, and then I will detach from this topic.

Reply:
> There is no concept of raising or lowering, there are just candidates with (default) priorities.

ah -- then there should be no need for any config changes? just RF code (if still needed) so priority for RIA remotes is higher and be done?

> I suspect that you want to aim for a bigger solution. If you insist, ...

oh no -- I am not insisting or even suggesting to implement anything bigger solution here.  I just want this solution not require changes to user visible parts (e.g. config) later if we decide to go with some global user visible approach.

Reply:
If an existing unprioritized candidate configuration is found, it automatically gets priority 700 such that identical behavior is maintained. This doesn't fix gh-4613, but I see no reason for doing such intervention in repos where this wasn't a problem so far (and I believe there are just a handful). All new clones will get the new behavior.